File: srclab-common/src/main/java/xyz/srclab/common/base/Check.java
package xyz.srclab.common.base;

import java.util.NoSuchElementException;

/**
 * @author sunqian
 */
public class Check {

    public static void checkArguments(boolean expression) {
        if (!expression) {
            throw new IllegalArgumentException();
        }
    }

    public static void checkArguments(boolean expression, Object message) {
        if (!expression) {
            throw new IllegalArgumentException(message);
        }
    }

    public static void checkArguments(boolean expression, String messagePattern, Object... messageArgs) {
        if (!expression) {
            throw new IllegalArgumentException(Format.fastFormat(messagePattern, messageArgs));
        }
    }

    public static void checkState(boolean expression) {
        if (!expression) {
            throw new IllegalStateException();
        }
    }

    public static void checkState(boolean expression, Object message) {
        if (!expression) {
            throw new IllegalStateException(message);
        }
    }

    public static void checkState(boolean expression, String messagePattern, Object... messageArgs) {
        if (!expression) {
            throw new IllegalStateException(Format.fastFormat(messagePattern, messageArgs));
        }
    }

    public static void checkNull(boolean expression) {
        if (!expression) {
            throw new NullPointerException();
        }
    }

    public static void checkNull(boolean expression, Object message) {
        if (!expression) {
            throw new NullPointerException(message);
        }
    }

    public static void checkNull(boolean expression, String messagePattern, Object... messageArgs) {
        if (!expression) {
            throw new NullPointerException(Format.fastFormat(messagePattern, messageArgs));
        }
    }

    public static void checkSupported(boolean expression) {
        if (!expression) {
            throw new UnsupportedOperationException();
        }
    }

    public static void checkSupported(boolean expression, Object message) {
        if (!expression) {
            throw new UnsupportedOperationException(message);
        }
    }

    public static void checkSupported(boolean expression, String messagePattern, Object... messageArgs) {
        if (!expression) {
            throw new UnsupportedOperationException(Format.fastFormat(messagePattern, messageArgs));
        }
    }

    public static void checkElement(boolean expression) {
        if (!expression) {
            throw new NoSuchElementException();
        }
    }

    public static void checkElement(boolean expression, Object key) {
        if (!expression) {
            throw new NoSuchElementException(ToString.toString(key));
        }
    }

    public static void checkElement(boolean expression, String messagePattern, Object... messageArgs) {
        if (!expression) {
            throw new NoSuchElementException(Format.fastFormat(messagePattern, messageArgs));
        }
    }

    public static void checkIndex(boolean expression) {
        if (!expression) {
            throw new IndexOutOfBoundsException();
        }
    }

    public static void checkIndex(boolean expression, Object index) {
        if (!expression) {
            throw new IndexOutOfBoundsException(ToString.toString(index));
        }
    }

    public static void checkIndex(boolean expression, String messagePattern, Object... messageArgs) {
        if (!expression) {
            throw new IndexOutOfBoundsException(Format.fastFormat(messagePattern, messageArgs));
        }
    }

    public static void checkIndexRange(int start, int end, int length) {
        if (start > end) {
            throw new IllegalArgumentException(
                    "Index range error: start > end [start : " + start + ", end: " + end + "]");
        }
        if (start < 0) {
            throw new IndexOutOfBoundsException("Index range error: start < 0 [start: " + start + "]");
        }
        if (end > length) {
            throw new IndexOutOfBoundsException(
                    "Index range error: end > length [end: " + end + ", length: " + length + "]");
        }
    }

    public static void checkIndexRange(int start, int end, int startBound, int endBound) {
        if (startBound > endBound) {
            throw new IllegalArgumentException(
                    "Range error: startBound > endBound " +
                            "[startBound : " + startBound + ", endBound: " + endBound + "]");
        }
        if (start > end) {
            throw new IndexOutOfBoundsException(
                    "Index range error: start > end [start : " + start + ", end: " + end + "]");
        }
        if (start < startBound) {
            throw new IndexOutOfBoundsException(
                    "Index range error: start < startBound " +
                            "[start: " + start + ", startBound: " + startBound + "]");
        }
        if (end > endBound) {
            throw new IndexOutOfBoundsException(
                    "Index range error: end > endBound " +
                            "[end: " + end + ", endBound: " + endBound + "]");
        }
    }

    public static void checkIndexIn(int index, int startBound, int endBound) {
        if (startBound > endBound) {
            throw new IllegalArgumentException(
                    "Range error: startBound > endBound " +
                            "[startBound : " + startBound + ", endBound: " + endBound + "]");
        }
        if (index < startBound || index > endBound) {
            throw new IndexOutOfBoundsException("Index: " + index +
                    "[startBound: " + startBound + ", endBound: " + endBound + "]");
        }
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/base/Check.kt
@file:JvmName("Check")
@file:JvmMultifileClass

package xyz.srclab.common.base

import java.util.*

fun checkArgument(expression: Boolean) {
    if (!expression) {
        throw IllegalArgumentException()
    }
}

fun checkArgument(expression: Boolean, message: String?) {
    if (!expression) {
        throw IllegalArgumentException(message)
    }
}

fun checkArgument(expression: Boolean, messagePattern: String?, vararg messageArgs: Any?) {
    if (!expression) {
        throw IllegalArgumentException(format(messagePattern, messageArgs))
    }
}

fun checkState(expression: Boolean) {
    if (!expression) {
        throw IllegalStateException()
    }
}

fun checkState(expression: Boolean, message: String?) {
    if (!expression) {
        throw IllegalStateException(message)
    }
}

fun checkState(expression: Boolean, messagePattern: String?, vararg messageArgs: Any?) {
    if (!expression) {
        throw IllegalStateException(format(messagePattern, messageArgs))
    }
}

fun checkNull(expression: Boolean) {
    if (!expression) {
        throw NullPointerException()
    }
}

fun checkNull(expression: Boolean, message: String?) {
    if (!expression) {
        throw NullPointerException(message)
    }
}

fun checkNull(expression: Boolean, messagePattern: String?, vararg messageArgs: Any?) {
    if (!expression) {
        throw NullPointerException(format(messagePattern, messageArgs))
    }
}

fun checkSupported(expression: Boolean) {
    if (!expression) {
        throw UnsupportedOperationException()
    }
}

fun checkSupported(expression: Boolean, message: String?) {
    if (!expression) {
        throw UnsupportedOperationException(message)
    }
}

fun checkSupported(expression: Boolean, messagePattern: String?, vararg messageArgs: Any?) {
    if (!expression) {
        throw UnsupportedOperationException(format(messagePattern, messageArgs))
    }
}

fun checkElement(expression: Boolean) {
    if (!expression) {
        throw NoSuchElementException()
    }
}

fun checkElement(expression: Boolean, message: String?) {
    if (!expression) {
        throw NoSuchElementException(message)
    }
}

fun checkElement(expression: Boolean, messagePattern: String?, vararg messageArgs: Any?) {
    if (!expression) {
        throw NoSuchElementException(format(messagePattern, messageArgs))
    }
}

fun checkBounds(expression: Boolean) {
    if (!expression) {
        throw IndexOutOfBoundsException()
    }
}

fun checkBounds(expression: Boolean, message: String?) {
    if (!expression) {
        throw IndexOutOfBoundsException(message)
    }
}

fun checkBounds(expression: Boolean, messagePattern: String?, vararg messageArgs: Any?) {
    if (!expression) {
        throw IndexOutOfBoundsException(format(messagePattern, messageArgs))
    }
}

fun isIndexInBounds(index: Int, startInclusive: Int, endExclusive: Int): Boolean {
    return index in startInclusive until endExclusive
}

fun isIndexInBounds(index: Int, startInclusive: Long, endExclusive: Long): Boolean {
    return index in startInclusive until endExclusive
}

fun checkIndexInBounds(index: Int, startInclusive: Int, endExclusive: Int) {
    checkBounds(
        isIndexInBounds(index, startInclusive, endExclusive),
        "Index out of bounds[" +
                "index: $index, startInclusive: $startInclusive, endExclusive: $endExclusive" +
                "]."
    )
}

fun checkIndexInBounds(index: Int, startInclusive: Long, endExclusive: Long) {
    checkBounds(
        isIndexInBounds(index, startInclusive, endExclusive),
        "Index out of bounds[" +
                "index: $index, startInclusive: $startInclusive, endExclusive: $endExclusive" +
                "]."
    )
}

fun isRangeInLength(startInclusive: Int, endExclusive: Int, length: Int): Boolean {
    return startInclusive >= 0 && endExclusive <= length && startInclusive <= endExclusive
}

fun isRangeInLength(startInclusive: Long, endExclusive: Long, length: Long): Boolean {
    return startInclusive >= 0 && endExclusive <= length && startInclusive <= endExclusive
}

fun isRangeInBounds(
    startInclusive: Int, endExclusive: Int, startBoundInclusive: Int, endBoundExclusive: Int
): Boolean {
    return startInclusive >= startBoundInclusive
            && endExclusive <= endBoundExclusive
            && startInclusive <= endExclusive
}

fun isRangeInBounds(
    startInclusive: Long, endExclusive: Long, startBoundInclusive: Long, endBoundExclusive: Long
): Boolean {
    return startInclusive >= startBoundInclusive
            && endExclusive <= endBoundExclusive
            && startInclusive <= endExclusive
}

fun checkRangeInLength(startInclusive: Int, endExclusive: Int, length: Int) {
    checkBounds(
        isRangeInLength(startInclusive, endExclusive, length),
        "Range out of bounds[" +
                "startInclusive: $startInclusive, endExclusive: $endExclusive, length: $length" +
                "]."
    )
}

fun checkRangeInLength(startInclusive: Long, endExclusive: Long, length: Long) {
    checkBounds(
        isRangeInLength(startInclusive, endExclusive, length),
        "Range out of bounds[" +
                "startInclusive: $startInclusive, endExclusive: $endExclusive, length: $length" +
                "]."
    )
}

fun checkRangeInBounds(startInclusive: Int, endExclusive: Int, startBoundInclusive: Int, endBoundExclusive: Int) {
    checkBounds(
        isRangeInBounds(startInclusive, endExclusive, startBoundInclusive, endBoundExclusive),
        "Range out of bounds[" +
                "startInclusive: $startInclusive, endExclusive: $endExclusive, " +
                "startBoundInclusive: $startBoundInclusive, endBoundExclusive: $endBoundExclusive" +
                "]."
    )
}

fun checkRangeInBounds(
    startInclusive: Long,
    endExclusive: Long,
    startBoundInclusive: Long,
    endBoundExclusive: Long
) {
    checkBounds(
        isRangeInBounds(startInclusive, endExclusive, startBoundInclusive, endBoundExclusive),
        "Range out of bounds[" +
                "startInclusive: $startInclusive, endExclusive: $endExclusive, " +
                "startBoundInclusive: $startBoundInclusive, endBoundExclusive: $endBoundExclusive" +
                "]."
    )
}

private fun format(pattern: String?, vararg args: Any?): String? {
    if (pattern === null) {
        return null
    }
    return pattern.fastFormat(args)
}